### PR TITLE
[inductor] Add single pass "var_unnormalized" reduction_type

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102737
* __->__ #102486

This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng

### CODE REVIEW

**Comment by ngimel** [comment]
Should we modify ir.Reduction to have multiple outputs?
Or could we remedy this in at the codegen level, by using `tl.reduce` that can now do multiple reductions at once, thanks to you?

**Comment by peterbell10** [comment]
> Should we modify `ir.Reduction` to have multiple outputs?

I tried going in this direction at first, but it's kind of baked into the IR format that there is one output per loop. Changing that seemed to be quite a large effort. Thankfully, my benchmarks show that this is still a huge win even if we do recompute the mean because the memory bandwidth is used so much more efficiently.

> Or could we remedy this in at the codegen level, by using `tl.reduce` that can now do multiple reductions at once, thanks to you?

Yeah merging multiple reduction ops into a single `tl.reduce` call could be interesting to explore. I do think it would be nicer if triton could do this automatically though. I think there are trade-offs in terms of instruction scheduling and shared memory usage that are a lot harder to reason about from inductor's side.

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -3636,15 +3636,48 @@ def var_mean_(x, axis, correction, keepdim, return_mean):
...
+        return data * scale
+
+    return make_pointwise(scale_fn)(uncorrected_var)
 
 
 @register_lowering(aten.var_mean)
 def var_mean(x, axis=None, *, correction=None, keepdim=False):
-    return var_mean_(
-        x, axis=axis, correction=correction, keepdim=keepdim, return_mean=True
+    if torch.version.hip is not None:
```
Comment (Chillee) [inline torch/_inductor/lowering.py:3673]: torch.version.hip?

**Comment by ngimel** [comment]
I'm slightly concerned about the impact this might have on `var_mean` reduction across a small dimension that actually allows persistent reduction (like layer_norm with size 768-1024 or maybe even 2048, though we don't allow persistent reduction for 2048 yet) - in this case persistent "2 pass" reduction might be better because data is loaded only once, and (I think) it requires fewer registers. Can we have different lowerings based on the reduction dimension? Even with dynamic shapes ln reduction dimention should be known at compile time.

**Review by jgong5 — COMMENTED** [review]
Do you have performance benchmarking comparing the the numbers before and after the changes?

File: torch/_inductor/codegen/cpp_prefix.h [hunk 2]
```diff
@@ -12,16 +12,116 @@
...
+    acc.m2 + delta * new_delta,
+    new_weight
+  };
+  return result;
+}
+
+
+#if INDUCTOR_USE_VECTOR_TYPES()
+template <typename scalar_t>
+inline at::vec::Vectorized<scalar_t> vec_shuffle_down(at::vec::Vectorized<scalar_t> x, size_t n) {
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_prefix.h:87]: better with permute instruction?

**Comment by peterbell10** [comment]
Looking at the benchmark results, this is a significant win for some specific timm models with dynamic shapes. 

model | change in speedup
----------|---------------------------
adv_inception_v3    | 1.1318 → 1.2497
gluon_inception_v3 | 1.1366 → 1.2473
inception_v3             | 1.1333 → 1.2473
timm_efficientdet | 1.1249 → 1.2690

Presumably the static shape variants are using persistent reductions, so don't benefit from this change.

https://hud.pytorch.org/benchmark/timm_models/inductor_dynamic?startTime=Wed%2C%2007%20Jun%202023%2021%3A10%3A11%20GMT&stopTime=Wed%2C%2014%20Jun%202023%2021%3A10%3A11%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/peterbell10/562/head&lCommit=dda51adf26e627762b00049eb67b80b79eed1f70&rBranch=main&rCommit=593642d1d8d0edec05eaeeca38bb2b8be9b73639

**Comment by Chillee** [comment]
IIRC, the other setting where this seemed like it would be a significant win was for channels-last batch norm. That might also be worth spot checking?

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -49,6 +49,24 @@
...
+            torch.float64,
+            torch.int8,
+            torch.int16,
+            torch.int32,
+            torch.int64,
+            torch.uint8,
+        ]
+    },
+}
+
```
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:69]: This is `_prims_common.get_computation_dtype`.

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1380,6 +1402,36 @@ def final_argreduce(buffer, result_var, value, index):
                 """
                 )
                 final_argreduce(self.suffix, result_var, accumulator, accumulator_index)
+            elif reduction_type == "var_unnormalized":
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1405]: It'd be so nice if we could generate these from the `combine_fn`...

File: torch/_inductor/lowering.py [hunk 5]
```diff
@@ -3768,56 +3768,66 @@ def _validate_reduction_axis(x, axis):
...
-                assert all(index[i] == 0 for i in reduced_idx)
-                index = [index[i] for i in kept_idx]
-            assert len(index) == len(kept_idx)
-            new_index = [None] * (len(index) + len(reduction_index))
-            for idx, var in itertools.chain(
-                zip(kept_idx, index), zip(reduced_idx, reduction_index)
-            ):
-                new_index[idx] = var
-            return inner_loader(new_index)
+def _make_reduction_inner(x, *, axis, keepdims, dtype, override_return_dtype):
```
Comment (lezcano) [inline torch/_inductor/lowering.py:3771]: why are the diffs so bad :'(

Comment (lezcano) [reply torch/_inductor/lowering.py:3771]: Is this part all just code shuffling?

**Review by lezcano — APPROVED** [review]
Cool! I have not reviewed the C++ generation code though. @jgong5 could you have a look at that part?

File: torch/_inductor/lowering.py (hunk 5 above)
Comment (peterbell10) [reply torch/_inductor/lowering.py:3771]: The diff is much clearer if you ignore whitespace changes,
https://github.com/pytorch/pytorch/pull/102486/files?diff=unified&w=1

Comment (lezcano) [reply torch/_inductor/lowering.py:3771]: TIL!

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -921,56 +968,35 @@ def fn(index):
...
 
-        split_reduction = (
-            is_triton(device)
-            and reduction_type
-            not in {
-                "argmax",
-                "argmin",
-            }
-            and config.split_reductions
+        # triton doesn't support reduce to single element well, so break it up
```
Comment (lezcano) [inline torch/_inductor/ir.py:971]: nit. This comment is repeated a few lines below, and now it's sort of stale, better remove it.

**Review by jansel — APPROVED** [review]
